In [1]:
import pandas as pd

In [5]:
df = pd.read_csv("/content/tips.csv")
df

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [6]:
df.head() # previews first 5 rows of a DataFrame

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [9]:
df.tail() # previews last 3 rows of a DataFrame; by default 5 oota garxa but we can change that

total_bill   tip     sex smoker   day    time  size
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

In [10]:
df.sample(5) # randomly samples 5 rows

total_bill   tip     sex smoker   day    time  size
231       15.69  3.00    Male    Yes   Sat  Dinner     3
144       16.43  2.30  Female     No  Thur   Lunch     2
80        19.44  3.00    Male    Yes  Thur   Lunch     2
8         15.04  1.96    Male     No   Sun  Dinner     2
228       13.28  2.72    Male     No   Sat  Dinner     2

In [11]:
# mathi ko sample harek patak run garda naya aauxa same aaune banauna seed chainxa
# pandas ra scikit-learn ma seed lai random_state vanxa

In [12]:
df.sample(5, random_state=2024)

total_bill    tip     sex smoker   day    time  size
239       29.03   5.92    Male     No   Sat  Dinner     3
238       35.83   4.67  Female     No   Sat  Dinner     3
170       50.81  10.00    Male    Yes   Sat  Dinner     3
156       48.17   5.00    Male     No   Sun  Dinner     6
134       18.26   3.25  Female     No  Thur   Lunch     2

In [13]:
df.shape # (no_of_rows, no_of_columns)

(244, 7)

# Data Profiling and inspection

In [14]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

In [16]:
df.head(3)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3

In [18]:
df.describe() # describes the numerical data

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

In [19]:
# 25 % => Q1
# 50 % => Q2, Median
# 75 % => Q3

In [20]:
df.dtypes

total_bill    float64
tip           float64
sex            object
smoker         object
day            object
time           object
size            int64
dtype: object

In [21]:
df.describe(include="object")

sex smoker  day    time
count    244    244  244     244
unique     2      2    4       2
top     Male     No  Sat  Dinner
freq     157    151   87     176

**sex column**
* 244 rows
* 2 unique categories
* among 2 unique categories most occuring is Male
* with frequency 157

In [22]:
df.sex.value_counts()

Male      157
Female     87
Name: sex, dtype: int64

In [24]:
df.sex.unique()

array(['Female', 'Male'], dtype=object)

In [25]:
df.sex.nunique()

2

In [26]:
df.time.value_counts()

Dinner    176
Lunch      68
Name: time, dtype: int64

In [28]:
df.day.value_counts()

Sat     87
Sun     76
Thur    62
Fri     19
Name: day, dtype: int64

In [30]:
df["smoker"].value_counts()

No     151
Yes     93
Name: smoker, dtype: int64

In [31]:
df.count()

total_bill    244
tip           244
sex           244
smoker        244
day           244
time          244
size          244
dtype: int64

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB


In [33]:
# null means missing data

In [37]:
df.isna() # checks if the data contains missing values or not

total_bill    tip    sex  smoker    day   time   size
0         False  False  False   False  False  False  False
1         False  False  False   False  False  False  False
2         False  False  False   False  False  False  False
3         False  False  False   False  False  False  False
4         False  False  False   False  False  False  False
..          ...    ...    ...     ...    ...    ...    ...
239       False  False  False   False  False  False  False
240       False  False  False   False  False  False  False
241       False  False  False   False  False  False  False
242       False  False  False   False  False  False  False
243       False  False  False   False  False  False  False

[244 rows x 7 columns]

In [36]:
# I want to count no. of missing values in each column

In [38]:
True + True

2

In [39]:
True + False

1

In [40]:
False + True

1

In [41]:
False + False

0

In [43]:
df.isna().sum()

total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

# loc vs iloc

In [44]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

# slice data with Male in sex column

In [48]:
df["sex"] == "Male"

0      False
1       True
2       True
3       True
4      False
       ...  
239     True
240    False
241     True
242     True
243    False
Name: sex, Length: 244, dtype: bool

In [51]:
df.loc[df["sex"] == "Male"]

total_bill   tip   sex smoker  day    time  size
1         10.34  1.66  Male     No  Sun  Dinner     3
2         21.01  3.50  Male     No  Sun  Dinner     3
3         23.68  3.31  Male     No  Sun  Dinner     2
5         25.29  4.71  Male     No  Sun  Dinner     4
6          8.77  2.00  Male     No  Sun  Dinner     2
..          ...   ...   ...    ...  ...     ...   ...
236       12.60  1.00  Male    Yes  Sat  Dinner     2
237       32.83  1.17  Male    Yes  Sat  Dinner     2
239       29.03  5.92  Male     No  Sat  Dinner     3
241       22.67  2.00  Male    Yes  Sat  Dinner     2
242       17.82  1.75  Male     No  Sat  Dinner     2

[157 rows x 7 columns]

In [52]:
df["sex"] == "Female"

0       True
1      False
2      False
3      False
4       True
       ...  
239    False
240     True
241    False
242    False
243     True
Name: sex, Length: 244, dtype: bool

In [55]:
df.loc[df["sex"] == "Female"]

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
11        35.26  5.00  Female     No   Sun  Dinner     4
14        14.83  3.02  Female     No   Sun  Dinner     2
16        10.33  1.67  Female     No   Sun  Dinner     3
..          ...   ...     ...    ...   ...     ...   ...
226       10.09  2.00  Female    Yes   Fri   Lunch     2
229       22.12  2.88  Female    Yes   Sat  Dinner     2
238       35.83  4.67  Female     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[87 rows x 7 columns]

# Who gives more tips on average Male or Female?

In [63]:
df.loc[df["sex"] == "Male"]["tip"].mean()

3.0896178343949043

In [64]:
df.loc[df["sex"] == "Female"]["tip"].mean()

2.8334482758620685

In [66]:
df.loc[df["day"] =="Fri"]

total_bill   tip     sex smoker  day    time  size
90        28.97  3.00    Male    Yes  Fri  Dinner     2
91        22.49  3.50    Male     No  Fri  Dinner     2
92         5.75  1.00  Female    Yes  Fri  Dinner     2
93        16.32  4.30  Female    Yes  Fri  Dinner     2
94        22.75  3.25  Female     No  Fri  Dinner     2
95        40.17  4.73    Male    Yes  Fri  Dinner     4
96        27.28  4.00    Male    Yes  Fri  Dinner     2
97        12.03  1.50    Male    Yes  Fri  Dinner     2
98        21.01  3.00    Male    Yes  Fri  Dinner     2
99        12.46  1.50    Male     No  Fri  Dinner     2
100       11.35  2.50  Female    Yes  Fri  Dinner     2
101       15.38  3.00  Female    Yes  Fri  Dinner     2
220       12.16  2.20    Male    Yes  Fri   Lunch     2
221       13.42  3.48  Female    Yes  Fri   Lunch     2
222        8.58  1.92    Male    Yes  Fri   Lunch     1
223       15.98  3.00  Female     No  Fri   Lunch     3
224       13.42  1.58    Male    Yes  Fri   Lunch     2
225       16.27  2.50  Female    Yes  Fri   Lunch     2
226       10.09  2.00  Female    Yes  Fri   Lunch     2

In [67]:
df.day.unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

# kun bar on average kati tip hunxa; sab vanda badi tip kun bar dinxa?

In [68]:
df.loc[df["day"] == "Sun"].tip.mean()

3.2551315789473687

In [69]:
df.loc[df["day"] == "Fri"].tip.mean()

2.7347368421052627

In [70]:
# yo ta each category lai garnu paryo; if categories are many its difficult to do this
# so yesko solution groupby ho

# groupby

In [71]:
# day ko adhar ma data lai group garnu paryo

In [73]:
df.groupby(["day"])["tip"].mean()

day
Fri     2.734737
Sat     2.993103
Sun     3.255132
Thur    2.771452
Name: tip, dtype: float64

In [74]:
df.groupby(["day"]).tip.mean()

day
Fri     2.734737
Sat     2.993103
Sun     3.255132
Thur    2.771452
Name: tip, dtype: float64

In [75]:
df.groupby(["sex", "smoker"])["tip"].mean()

sex     smoker
Female  No        2.773519
        Yes       2.931515
Male    No        3.113402
        Yes       3.051167
Name: tip, dtype: float64

# loc vs iloc

In [78]:
# loc

In [79]:
# I want rows 14, 15, 16 and all columns

In [81]:
df.loc[14:16, :] # loc use garda ro slice garda last ko index linxa eg 14:16 means 14, 15, 16

total_bill   tip     sex smoker  day    time  size
14       14.83  3.02  Female     No  Sun  Dinner     2
15       21.58  3.92    Male     No  Sun  Dinner     2
16       10.33  1.67  Female     No  Sun  Dinner     3

In [82]:
# row 14-16 and column tip, sex

In [83]:
df.loc[14:16, ["tip", "sex"]]

tip     sex
14  3.02  Female
15  3.92    Male
16  1.67  Female

In [86]:
df.loc[13:16, "tip":"day"]

tip     sex smoker  day
13  3.00    Male     No  Sun
14  3.02  Female     No  Sun
15  3.92    Male     No  Sun
16  1.67  Female     No  Sun

In [87]:
# kasailai tip ra sex column ta index 1 and 2 ma xa why not to use index instead of column name
# lagna sakxa

In [90]:
# df.loc[14:16, 1:2] # error raise garxa

# yedi column ko index diney vane iloc garnu parxa

In [91]:
df.iloc[12:14, 1:3]

tip   sex
12  1.57  Male
13  3.00  Male

# How to slice column

In [92]:
# I want sabai rows tip and total_bill column

In [94]:
df[["tip", "total_bill"]]

tip  total_bill
0    1.01       16.99
1    1.66       10.34
2    3.50       21.01
3    3.31       23.68
4    3.61       24.59
..    ...         ...
239  5.92       29.03
240  2.00       27.18
241  2.00       22.67
242  1.75       17.82
243  3.00       18.78

[244 rows x 2 columns]

In [95]:
df.loc[:, ["tip", "total_bill"]]

tip  total_bill
0    1.01       16.99
1    1.66       10.34
2    3.50       21.01
3    3.31       23.68
4    3.61       24.59
..    ...         ...
239  5.92       29.03
240  2.00       27.18
241  2.00       22.67
242  1.75       17.82
243  3.00       18.78

[244 rows x 2 columns]